### ISSUE TITLE
Document AMP startup and shutdown APIs

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3476]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3476|
|Type|enhancement|
|Created|2008-10-16 14:51:12Z|
|Branch|https://github.com/twisted/twisted/tree/amp-docs-3476|

While `twisted.protocols.amp.AMP` is an `IProtocol` implementation, it adds its own layer of protocol startup and shutdown APIs on top of `makeConnection` and `connectionLost`: `startReceivingBoxes` and `stopReceivingBoxes`.

Any `IBoxReceiver`-handling code which attempts to follow the defined interfaces must call `startReceivingBoxes` and `stopReceivingBoxes`, not `makeConnection` and `connectionLost`.  This is because the latter two methods aren't part of `IBoxReceiver` and because there isn't necessarily an `ITCPTransport` provider to pass to `makeConnection`.

This means that any `AMP` subclass which is really being used as an `IBoxReceiver` implementation (subclassing `AMP` is the simplest way to make such a thing, at least if you want the whole command-handling stack) should be careful to avoid putting any application logic in `makeConnection` (or `connectionMade`) or `connectionLost`.  If the class ends up being handled by any code that expects `IBoxReceiver` providers, those methods won't be called.  Applications should instead override `startReceivingBoxes` and `stopReceivingBoxes` (also being careful to call the base implementations, since they do important stuff).

This needs to be documented, since it's a divergence from what one usually does when subclassing an `IProtocol` implementation, and `AMP` inherits methods from `BaseProtocol` with docstrings that claim you should override them when actually you shouldn't.

<details><summary>Searchable metadata</summary>

```
trac-id__3476 3476
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_amp_docs_3476 branches/amp-docs-3476
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1224168672000000 1224168672000000
changetime__1315315069000000 1315315069000000
version__None None
owner__exarkun exarkun
cc__thijs cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Unfortunately, if you implement things the "right way" in this case, your application won't work, due to a bug.  I've filed [#3477](https://github.com/twisted/twisted/issues/3477), which should probably be a prerequisite.  (Luckily it's pretty trivial to fix.)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [25248]) Merge early-sendbox-3477

Author: exarkun
Reviewer: glyph
Fixes: [#3477](https://github.com/twisted/twisted/issues/3477)
Refs: [#3476](https://github.com/twisted/twisted/issues/3476)


Fix `BinaryBoxProtocol` so that its `IBoxReceiver` can send boxes in its
`startReceivingBoxes` callback such that they are actually written to the
transport rather than result in an exception.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

Where should this be documented, in a howto or docstring?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Howto please.  There's already reasonable docstring coverage for all the relevant APIs.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [32491]) Branching to 'amp-docs-3476'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Not the most inspired document ever, but it explains some things.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/amp-docs-3476)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess set owner to @exarkun|
|-|-|

Thanks for seeing this old documentation through, exarkun. A few things:

 *  the howto isn't listed in the core howto index
 * "its arguments, respose, and possible error conditions" ==> "its arguments, response, and possible error conditions"
 * in the first code snippet in the Commands section, UsernameUnavailable is used before it is defined

Other than that, looks great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [32506]) Add AMP doc to the index

refs [#3476](https://github.com/twisted/twisted/issues/3476)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [32507]) fix typo in "response"

refs [#3476](https://github.com/twisted/twisted/issues/3476)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [32508]) Define UsernameUnavailable in the code snippets that use it

refs [#3476](https://github.com/twisted/twisted/issues/3476)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [32511]) Merge amp-docs-3476 

Author: exarkun
Reviewer: jesstess
Fixes: [#3476](https://github.com/twisted/twisted/issues/3476)

Add a howto covering some of the basics of AMP.
